pkg/kv/dynamodb/store.go

```diff
@@ -0,0 +1,320 @@
+		},
+	})
+	if err != nil {
+		if _, ok := err.(*dynamodb.ResourceInUseException); !ok {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Store) bytesKeyToDynamoKey(key []byte) map[string]*dynamodb.AttributeValue {
+	return map[string]*dynamodb.AttributeValue{
+		PartKey: {
+			S: aws.String(s.getPartitionKey(key)),
+		},
+		ItemKey: {
+			S: aws.String(string(key)),
+		},
+	}
+}
+
+func (s *Store) Get(ctx context.Context, key []byte) (*kv.ValueWithPredicate, error) {
+	if len(key) == 0 {
+		return nil, kv.ErrMissingKey
+	}
+
+	result, err := s.svc.GetItemWithContext(ctx, &dynamodb.GetItemInput{
+		TableName: aws.String(s.params.TableName),
+		Key:       s.bytesKeyToDynamoKey(key),
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if result.Item == nil {
+		return nil, kv.ErrNotFound
+	}
+
+	var item DynKVItem
+	err = dynamodbattribute.UnmarshalMap(result.Item, &item)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+
+	return &kv.ValueWithPredicate{
+		Value:     []byte(item.ItemValue),
+		Predicate: kv.Predicate([]byte(item.ItemValue)),
+	}, nil
+}
+
+func (s *Store) Set(ctx context.Context, key, value []byte) error {
+	return s.setWithOptionalPredicate(ctx, key, value, nil, false)
+}
+
+func (s *Store) SetIf(ctx context.Context, key, value []byte, valuePredicate kv.Predicate) error {
+	return s.setWithOptionalPredicate(ctx, key, value, valuePredicate, true)
+}
+
+func (s *Store) setWithOptionalPredicate(ctx context.Context, key, value []byte, valuePredicate kv.Predicate, usePredicate bool) error {
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+
+	item := DynKVItem{
+		PartKey:   s.getPartitionKey(key),
+		ItemKey:   string(key),
+		ItemValue: string(value),
+	}
+
+	marshaledItem, err := dynamodbattribute.MarshalMap(item)
+	if err != nil {
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:      marshaledItem,
+		TableName: &s.params.TableName,
+	}
+	if usePredicate {
+		if valuePredicate != nil {
+			input.ConditionExpression = aws.String(ItemValue + " = :predicate")
+			input.ExpressionAttributeValues = map[string]*dynamodb.AttributeValue{
+				":predicate": {S: aws.String(string(valuePredicate.([]byte)))},
+			}
+		} else {
+			input.ConditionExpression = aws.String("attribute_not_exists(" + ItemValue + ")")
+		}
+	}
+
+	_, err = s.svc.PutItemWithContext(ctx, input)
+
+	if err != nil {
+		if _, ok := err.(*dynamodb.ConditionalCheckFailedException); ok && usePredicate {
+			return kv.ErrPredicateFailed
+		}
+		return fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
+	}
+	return nil
+}
+
+func (s *Store) Delete(ctx context.Context, key []byte) error {
+	if len(key) == 0 {
+		return kv.ErrMissingKey
+	}
+
+	_, err := s.svc.DeleteItemWithContext(ctx, &dynamodb.DeleteItemInput{
+		TableName: aws.String(s.params.TableName),
+		Key:       s.bytesKeyToDynamoKey(key),
+	})
+
+	if err != nil {
```

Comment:
Does DynamoDB return ErrNotFound if the key is missing?

Reply:
Nope. `DeleteItem` (and `DeleteItemWithContext` for that matter) is idempotent and as such does not fails if the key (and, thus, the item) does not exist. That aligns with our `KV` interface